### PR TITLE
util: add var int codecs

### PR DESCRIPTION
**Pull request by BusyJay**
@ngaut @coocood @qiuyesuifeng PTAL

### CODE REVIEW

File: src/util/codec/mod.rs [hunk 1]
```diff
@@ -39,6 +39,13 @@ quick_error! {
             description("out of bound.")
             display("want {} actual {}", want, actual)
         }
+        InvalidDataType(reason: String) {
+            description("invalid data type")
+            display("{}", reason)
+        }
+        Eof {
```
Comment (qiuyesuifeng) [inline src/util/codec/mod.rs]: Should we use `EOF` instead?

Comment (BusyJay) [inline src/util/codec/mod.rs]: No. Only first character need to be upper case.

File: src/util/codec/number.rs [hunk 2]
```diff
@@ -70,29 +71,120 @@ pub fn encode_u64_desc(buf: &mut [u8], v: u64) -> Result<()> {
...
+    i + 1
+}
+
+/// `decode_var_u64` decodes value encoded by `encode_var_u64` before.
+pub fn decode_var_u64(buf: &[u8]) -> Result<(u64, usize)> {
+    let (mut x, mut s) = (0, 0);
+    for (i, &b) in buf.iter().enumerate() {
+        if b < 0x80 {
+            if i > 9 || i == 9 && b > 1 {
+                return Err(Error::OutOfBound(8, i));
```
Comment (qiuyesuifeng) [inline src/util/codec/number.rs]: Seems this error is not out of bound, but is an invalid encode one?

File: src/util/codec/number.rs [hunk 3]
```diff
@@ -110,38 +202,103 @@ mod test {
...
         ordered_case.reverse();
-        test_order(&test_func[1], &test_values, &ordered_case);
+        test_order(&test_func[1], tests, &ordered_case);
+    }
+
+    #[test]
+    fn test_var_i64_codec() {
+        for &v in I64_TESTS {
+            let mut buf = vec![0; 10];
+            assert!(encode_var_i64(&mut buf, v) <= 10);
```
Comment (qiuyesuifeng) [inline src/util/codec/number.rs]: Should use MAX_VAR_I64_LEN instead?

File: src/util/codec/number.rs [hunk 4]
```diff
@@ -110,38 +202,103 @@ mod test {
...
+
+        buf = vec![0x80; 9];
+        buf.push(0x2);
+        test_decode_out_of_bound(decode_var_u64, &buf);
+        test_decode_out_of_bound(decode_var_i64, &buf);
+
+        buf = vec![0x80; 3];
+        match decode_var_u64(&buf) {
+            Err(Error::Eof) => {}
+            o => panic!("eof is epxpected, but we got: {:?}", o),
```
Comment (qiuyesuifeng) [inline src/util/codec/number.rs]: expected

File: src/util/codec/number.rs (hunk 2 above)
Comment (BusyJay) [inline src/util/codec/number.rs]: If we reach here, it's overflow.

**Comment by qiuyesuifeng** [comment]
LGTM

**Comment by BusyJay** [comment]
@ngaut @coocood PTAL

**Comment by coocood** [comment]
LGTM

**Comment by ngaut** [comment]
LGTM

**Pull request merged**